### PR TITLE
complete feature compliance checking and add documentation

### PR DESCRIPTION
**Pull request by ckrieger**
…ology modeler

Signed-off-by: Christoph Krieger <krieger.christoph1@gmail.com>

- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request. Especially, we require **a single commit**
- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Ensure to use auto format in **all** files
- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for UI changes)

### CODE REVIEW

File: org.eclipse.winery.compliance/src/main/java/org/eclipse/winery/compliance/matching/TOSCAComplianceRuleMatcher.java [hunk 1]
```diff
@@ -30,88 +30,88 @@
 
 public class TOSCAComplianceRuleMatcher implements ITOSCAMatcher {
```
Comment (lharzenetter) [inline org.eclipse.winery.compliance/src/main/java/org/eclipse/winery/compliance/matching/TOSCAComplianceRuleMatcher.java]: Our naming convention is that we always use camel case - even for abbreveations to enhance readability. So Please rename this class to `ToscaComplianceRuleMatcher`.

In this case, is it possible to also rename all other classes which do not follow this convention in the compliance project? (`ITOSCAMatcher`, `TOSCANode`, ...)

File: org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/ServiceTemplateResource.java [hunk 2]
```diff
@@ -292,10 +292,10 @@ public Response injectNodeTemplates(InjectorReplaceData injectorReplaceData, @Co
     @Path("constraintchecking/")
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.rest/src/main/java/org/eclipse/winery/repository/rest/resources/servicetemplates/ServiceTemplateResource.java]: Please remove the trailing slash.

File: docs/user/ComplianceChecking.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+An example of a valid compliance rule is shown here:
+
+![complianceRule.PNG](graphics/compliance-checking/complianceRule.PNG)
+
+The Identifier consists of a single Node Template with a property "DataType: PersonalData". As the Required Structure contains a Node Template of the same type and with a property with equivalent key and value it exists a matching mapping.
+Hence, the compliance rule is valid.
+
+## Concept of Compliance Checking
+The detection of relevant compliance areas in Topology Models is based on the detection of subgraph isomorphisms that also consider the types and attributes of the Topology Elements that form the Topology Model.
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: ... of subgraph isomorphism which also takes the types and attributes of Topology Elements forming the Topology Model into account.

File: docs/user/ComplianceChecking.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+An example of a compliance rule that is **detected** and **satisfied** for a Topology Template is shown here:
+
+![complianceRulesDetection.png](graphics/compliance-checking/complianceRuleDetection.PNG)
+
+**Detected** means: A compliance rule is detected if there is at least on matching mapping from the Identifier to the Topology Model
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: Add a comma infront of the `if` and add a full stop please.

`A compliance rule is detected, if there is at least one matching mapping between the Identifier and the Topology Model.`

File: docs/user/ComplianceChecking.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+**Satisfied** means: A compliance Rule is satisfied if each area that has been found by the detector also satisfies the Required Structure. Otherwise a rule is unsatisfied.
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: lowercase rule
add a comma before the if

File: docs/user/ComplianceChecking.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+ 
+
+## Tutorial
+In this section we will give you a simple example on how to use the topology compliance checking.
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: `In the following, a simple example is given describing how to use Topology Compliance checking.`

File: docs/user/ComplianceChecking.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+We will describe how to create your first topology-based compliance rule and how to check the compliance of topology templates based on existing compliance rules.
+
+Note: You should have a look at the [[Winery User Guide]](http://eclipse.github.io/winery/UserGuide) before reading this tutorial to get all the basic knowledge about winery.
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: Winery in uppercase

File: docs/user/ComplianceChecking.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+
+
+**Step 1:** If you need help getting winery up and running have a look at the [[Quickstart Guide]](http://eclipse.github.io/winery/user/quickstart.html)
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: Maybe use a third-level heading for this (``### Step 1: ...`).

File: docs/user/ComplianceChecking.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+**Step 2:** Create a compliance rule 
+
+First we need to create our first compliance rule.
+Select the tab  "Other Elements" => "Compliance Rules" to get an overview of the existing compliance rules in the Repository.
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: repository

File: docs/user/ComplianceChecking.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+Click on "Add new" to create a new compliance rule. A modal to add a new Compliance Rule appears: 
+
+![createComplianceRule.PNG](graphics/compliance-checking/createComplianceRule.PNG)
+
+1. Enter a name for your rule e.g "PersonalDataComplianceRule"
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: e.g.,

File: docs/user/ComplianceChecking.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+
+ 
+
+
+
+
+## License
+
+Copyright (c) 2013-2018 Contributors to the Eclipse Foundation
```
Comment (lharzenetter) [inline docs/user/ComplianceChecking.md]: This file was created recently: the year is only 2018.

Comment (ckrieger) [reply docs/user/ComplianceChecking.md]: thanks for the comments i will fix them :)

File: docs/user/ComplianceChecking.md (hunk 4 above)
Comment (ckrieger) [reply docs/user/ComplianceChecking.md]: i think a comma is not needed check this link: https://english.stackexchange.com/questions/37856/do-you-need-a-comma-before-a-subordinating-conjunction-like-if

File: docs/user/ComplianceChecking.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
+# Topology Compliance Checking  
+The Topology Compliance Checking of Winery enables to describe restrictions, constraints, and requirements for Topology Templates in form of reusable topology-based Compliance Rules.
```
Comment (koppor) [inline docs/user/ComplianceChecking.md]: Please add empty line after headings. Refs https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint.

File: docs/user/ComplianceChecking.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+The detection of relevant compliance areas in Topology Models is based on the detection of subgraph isomorphisms which also takes the types and attributes of Topology Elements forming the Topology Model into account.
+An example of a Compliance Rule that is **detected** and **satisfied** for a Topology Template is shown here:
+
+![complianceRulesDetection.png](graphics/compliance-checking/complianceRuleDetection.PNG)
+
+**Detected** means: A Compliance Rule is detected if there is at least on matching mapping from the Identifier to the Topology Model.
+
+**Satisfied** means: A Compliance Rule is satisfied if each area that has been found by the detector also satisfies the Required Structure. Otherwise a rule is unsatisfied.
+ 
+
```
Comment (koppor) [inline docs/user/ComplianceChecking.md]: not two blank lines

File: docs/user/ComplianceChecking.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+An exemplary Compliance Rule that is concerned with the storage of personal user data may look like this:
+
+Identifier             |  Required Structure
+:-------------------------:|:-------------------------:
+![identifier_modeler.PNG](graphics/compliance-checking/identifier_modeler.PNG)  |  ![required-structure.PNG](graphics/compliance-checking/required-structure.PNG)
+
+The motivation for this Compliance Rule is to ensure that all MySQL databases that store personal data are hosted on a specific OpenStack Component with the HypervisorEndpoint: "192.168.4.3" .
+
+
```
Comment (koppor) [inline docs/user/ComplianceChecking.md]: Not three blank lines

please check the wohle file.

File: docs/user/ComplianceChecking.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+2. Click "Open Editor"
+3. Model "Identifier" or "Required Structure"
+
+An exemplary Compliance Rule that is concerned with the storage of personal user data may look like this:
+
+Identifier             |  Required Structure
+:-------------------------:|:-------------------------:
+![identifier_modeler.PNG](graphics/compliance-checking/identifier_modeler.PNG)  |  ![required-structure.PNG](graphics/compliance-checking/required-structure.PNG)
+
+The motivation for this Compliance Rule is to ensure that all MySQL databases that store personal data are hosted on a specific OpenStack Component with the HypervisorEndpoint: "192.168.4.3" .
```
Comment (koppor) [inline docs/user/ComplianceChecking.md]: Remove space beteween " and .

File: docs/user/index.md [hunk 16]
```diff
@@ -32,7 +32,7 @@ The repository is also responsible for importing and  exporting CSARs, the excha
 - [Splitting](Splitting) - splitting functionality
 - [TopologyCompletion](TopologyCompletion) - topology completion with a [Tutorial](TopologyCompletionTutorial)
 - [XaaSPackager](XaaSPackager) - Enables reusing modeled topologies as templates for single applications
-
```
Comment (koppor) [inline docs/user/index.md]: Add empty line after

File: org.eclipse.winery.topologymodeler.ui/src/app/models/topologyModelerConfiguration.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+export class TopologyModelerConfiguration {
+    constructor(readonly id: string,
+                readonly ns: string,
+                readonly repositoryURL: string,
+                readonly uiURL: string,
+                readonly compareTo: string,
+                readonly parentPath = 'servicetemplates',
+                readonly elementPath = 'topologytemplate') {
+
```
Comment (koppor) [inline org.eclipse.winery.topologymodeler.ui/src/app/models/topologyModelerConfiguration.ts]: remove empty line

**Comment by koppor** [comment]
Filed 432cd835d6820290446811ef696c74e32eb4d558 as [CQ 16900](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=16900).

**Comment by lharzenetter** [comment]
Merge based on parallel IP

**Pull request merged**